File: python/119.杨辉三角-ii.py
#
# @lc app=leetcode.cn id=119 lang=python
#
# [119] 杨辉三角 II
#
# 解法1(T93% S100%): 主要要解决空间复杂度的问题，用一个目标行长度的队列来维护，每次取出首元素跟上次的数字相加塞到队尾，对于第i行循环i次即可，最后再多插入队尾一个1

# @lc code=start
class Solution(object):
    def getRow(self, rowIndex):
        """
        :type rowIndex: int
        :rtype: List[int]
        """
        ans = [1]
        for i in range(rowIndex):
            last = 0
            for j in range(i+1):
                first = ans.pop(0)
                ans.append(first + last)
                last = first
            ans.append(1)
        return ans
# @lc code=end

File: javascript/119.杨辉三角-ii.js
/*
 * @lc app=leetcode.cn id=119 lang=javascript
 *
 * [119] 杨辉三角 II
 */

// @lc code=start
/**
 * @param {number} rowIndex
 * @return {number[]}
 */
var getRow = function(rowIndex) {
    if (rowIndex == 0) { return [1]; }
    var queue = [1, 1];
    for (var i=2; i<=rowIndex; i++) {
        queue.push(1);
        var last = queue.shift();
        for (var j=1; j<i; j++) {
            var tmp = queue.shift();
            queue.push(last + tmp);
            last = tmp;
        }
        queue.push(1);
    }
    return queue;
};
// @lc code=end

